### PR TITLE
mgr/dashboard: performance counter browsers

### PR DESCRIPTION
**Pull request by Rubab-Syed**
http://tracker.ceph.com/issues/22521

This PR adds performance counter links for: osd, mon and rgw from their individual detail page as well as from "servers" page.

### CODE REVIEW

File: src/pybind/mgr/dashboard/module.py [hunk 1]
```diff
@@ -1102,7 +1139,8 @@ def _osd_summary(self, osd_id, osd_info):
                 result['up'] = osd_info['up']
                 result['in'] = osd_info['in']
 
-                result['url'] = get_prefixed_url("/osd/perf/{0}".format(osd_id))
+                #result['url'] = get_prefixed_url("/osd/perf/{0}".format(osd_id))
+                result['url'] = get_prefixed_url("/perf_counters/osd/{0}".format(osd_id))
```
Comment (jcsp) [inline src/pybind/mgr/dashboard/module.py]: This should still link to /perf/, and then the page at /perf/ should have a link to the /perf_counters/ equivalent.

File: src/pybind/mgr/dashboard/module.py [hunk 2]
```diff
@@ -826,6 +826,43 @@ def _servers(self):
...
+                        counter["unit"]  = "B/s"
+                    else:
+                        counter["value"] = global_instance().get_latest(service_type, service_id, key)
+                        counter["unit"] = ""
+                    counters.append(counter)
+                                                    
+                return {
+                    'service_type': service_type,
+                    'service_id': service_id,
+                    'schema': counters,
```
Comment (jcsp) [inline src/pybind/mgr/dashboard/module.py]: Now that this contains the data (not just the descriptions), let's not call it `schema`.  Just `counters` would probably make sense.

File: src/pybind/mgr/dashboard/module.py [hunk 3]
```diff
@@ -826,6 +826,43 @@ def _servers(self):
...
+                    content_data=json.dumps(self.perf_counters_data(service_type, service_id), indent=2)
+                )
+            
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
+            def perf_counters_data(self, service_type, service_id):
+                schema = global_instance().get_perf_schema(service_type, str(service_id)).values()[0]
+                counters = []
+                
+                for key, value in schema.items():
```
Comment (jcsp) [inline src/pybind/mgr/dashboard/module.py]: Let's sort the list of counters alphabetically at some stage in this process

File: src/pybind/mgr/dashboard/module.py [hunk 4]
```diff
@@ -826,6 +826,43 @@ def _servers(self):
...
+                    counter = dict()
+                    counter["name"] = str(key)
+                    counter["description"] = value["description"]
+                    if int(value["type"]) == 10:
+                        counter["value"] = global_instance().get_rate(service_type, service_id, key)
+                        counter["unit"]  = "B/s"
```
Comment (jcsp) [inline src/pybind/mgr/dashboard/module.py]: This should be `/s` rather than `B/s`, as not all the rate quantities are bytes

**Comment by Rubab-Syed** [comment]
perf0
perf1

**Comment by jcsp** [comment]
@Rubab-Syed this is good to merge once it's rebased on master (there's just a newline at the end of mgr_module.py that's confusing the merge).

**Pull request merged**